### ISSUE TITLE
package module fails to install any package in Fedora container due to missing python2-dnf

### ISSUE THREAD

**Issue by thozza** [ansible/ansible#22305]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
dnf module (via package module)

##### ANSIBLE VERSION

```
ansible 2.2.1.0
```

##### CONFIGURATION
No ansible.cfg is present on the system.

##### OS / ENVIRONMENT

I'm running CI for Ansible role in Travis CI using Test Kitchen with Docker as a backend. The testing playbook using the role is then run locally in a Fedora 25 Docker container provisioned by Test Kitchen.

FYI the [failed Travis CI build](https://travis-ci.org/thozza/ansible-gitlab-runner/builds/207980783#L924)

##### SUMMARY
Fedora docker images and cloud images (I'm not sure about the rest of the variants) don't contain `python2-dnf` package, which is required by the dnf ansible module. As a result, one is not able to install package on Fedora using package or dnf ansible modules.

The `python2-dnf` package should be installed by the dnf module if it is not available on the system [code](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/packaging/os/dnf.py#L174), but it does not seem to work as it ends on the error log above that.

##### STEPS TO REPRODUCE
I prepared minimal reproducer using Vagrant in [my repo](https://github.com/thozza/ansible-dnf-module-bug-reproducer)

Basically just run this playbook locally on Fedora system without python2-dnf installed (e.g. default Fedora Docker image or default Fedora cloud image.

```yaml
---
- hosts: localhost
  tasks:
  - name: Install tmux using package module
    package:
      name: tmux
      state: latest
    become: true
    ignore_errors: true

  - name: Install sl using dnf module
    dnf:
      name: sl
      state: latest
    become: true
```

##### EXPECTED RESULTS

It should just work and ideally python2-dnf should be installed in the background if needed.
```
[vagrant@localhost vagrant]$ ansible-playbook -i inventory -c local -v reproducer.yml
No config file found; using defaults

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
ok: [localhost]

TASK [Install tmux using package module] ***************************************
changed: [localhost] => {"changed": true, "results": ["Installed: libevent-2.0.22-1.fc25.x86_64", "Installed: tmux-2.2-3.fc25.x86_64"]}

TASK [Install sl using dnf module] *********************************************
changed: [localhost] => {"changed": true, "results": ["Installed: sl-5.02-3.fc24.x86_64"]}

PLAY RECAP *********************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=0   
```

##### ACTUAL RESULTS
```
[vagrant@localhost vagrant]$ ansible-playbook -i inventory -c local -vvvv reproducer.yml
No config file found; using defaults
Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/site-packages/ansible/plugins/callback/__init__.pyc

PLAYBOOK: reproducer.yml *******************************************************
1 plays in reproducer.yml

PLAY [localhost] ***************************************************************

TASK [setup] *******************************************************************
Using module file /usr/lib/python2.7/site-packages/ansible/modules/core/system/setup.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279 `" && echo ansible-tmp-1488793962.14-190159984716279="` echo ~/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279 `" ) && sleep 0'
<localhost> PUT /tmp/tmphH5MZo TO /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279/setup.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279/ /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279/setup.py && sleep 0'
<localhost> EXEC /bin/sh -c '/usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279/setup.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1488793962.14-190159984716279/" > /dev/null 2>&1 && sleep 0'
ok: [localhost]

TASK [Install tmux using package module] ***************************************
task path: /vagrant/reproducer.yml:4
Running dnf
Using module file /usr/lib/python2.7/site-packages/ansible/modules/extras/packaging/os/dnf.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871 `" && echo ansible-tmp-1488793962.47-106188174901871="` echo ~/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871 `" ) && sleep 0'
<localhost> PUT /tmp/tmppNYqOl TO /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871/dnf.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871/ /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871/dnf.py && sleep 0'
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-lgrufzejmmfmwxkmmrdubihjrxftmunw; /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871/dnf.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1488793962.47-106188174901871/" > /dev/null 2>&1'"'"' && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disablerepo": [], 
            "enablerepo": [], 
            "list": null, 
            "name": [
                "tmux"
            ], 
            "state": "latest"
        }
    }, 
    "msg": "`python2-dnf` is not installed, but it is required for the Ansible dnf module."
}
...ignoring

TASK [Install sl using dnf module] *********************************************
task path: /vagrant/reproducer.yml:11
Using module file /usr/lib/python2.7/site-packages/ansible/modules/extras/packaging/os/dnf.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vagrant
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo ~/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300 `" && echo ansible-tmp-1488793962.62-245842485197300="` echo ~/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300 `" ) && sleep 0'
<localhost> PUT /tmp/tmp8pdGBQ TO /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300/dnf.py
<localhost> EXEC /bin/sh -c 'chmod u+x /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300/ /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300/dnf.py && sleep 0'
<localhost> EXEC /bin/sh -c 'sudo -H -S -n -u root /bin/sh -c '"'"'echo BECOME-SUCCESS-vjtwhwrbvdohedhtuncgnrllfyycuyrb; /usr/bin/python /home/vagrant/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300/dnf.py; rm -rf "/home/vagrant/.ansible/tmp/ansible-tmp-1488793962.62-245842485197300/" > /dev/null 2>&1'"'"' && sleep 0'
fatal: [localhost]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "invocation": {
        "module_args": {
            "conf_file": null, 
            "disable_gpg_check": false, 
            "disablerepo": [], 
            "enablerepo": [], 
            "list": null, 
            "name": [
                "sl"
            ], 
            "state": "latest"
        }, 
        "module_name": "dnf"
    }, 
    "msg": "`python2-dnf` is not installed, but it is required for the Ansible dnf module."
}
	to retry, use: --limit @/vagrant/reproducer.retry

PLAY RECAP *********************************************************************
localhost                  : ok=2    changed=0    unreachable=0    failed=1   
```

**Comment by ansibot** [comment]
cc @ignatenkobrain @DJMuggs
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by igor-raits** [comment]
@abadger can you please look at this? I have no idea, why it falls into `module.check_mode` and what that means...

**Comment by abadger** [comment]
AFAICT, this is not a bug (and actually not related to check mode either).

If you want to use ansible and the dnf module with python2, then you need to look at this: https://fedoramagazine.org/getting-ansible-working-fedora-23/  Since Fedora doesn't install the packages the dnf module would need to function on python2, it doesn't work until you install those modules.  The Fedora magazine article describes how to get those packages installed.

If you want to use ansible and the dnf module with python3 then you need to configure your host to use python3 for modules.  Do that by setting ansible_python_interpreter=/path/to/python3 for that host.
(For instance, you could have this inventory entry to use python3 with localhost):

```
localhost ansible_python_interpreter=/usr/bin/python3

# Alternate method using a group to specify all fedora hosts which should use python3
[fedora]
localhost
fedora25.lan

[fedora:vars]
ansible_python_interpreter=/usr/bin/python3
```
Note that not all modules may be compatible with python3 at this time but we will treat incompatibilities as bugs to be fixed in subsequent releases.

**Comment by thozza** [comment]
@abadger I think it is pretty reasonable to expect ansible to be able to install packages using standard module on Fedora system without additional workarounds. It seems to me as pretty basic requirement from a management system like ansible. While you noted bunch of links and information you didn't answer why the explicit installation of python2-dnf in the module does not work. Authors seem to expect this situation and rather than requiring people to use workarounds that don't make much sense, they added those lines of code to install python2-dnf if it is not installed. It seems to me that it does not work and it should.

**I'm not trying to do any fancy thing and not trying to use Ansible with Python3.** I installed Python 2 and then installed Ansible using pip. It still does not work. Even if I forget that DNF is used on Fedora, I would at least expect the standard package module to work regardless what is on the system. Why you can not fall back to call dnf command directly? So instead of me being able to use this one generic module for almost ANY OS, I have to now deploy custom commands just for Fedora. To me it looks like Ansible didn't pull in all of its dependencies from PyPI to work with its core modules which it itself installs.

This is a bug from Ansible user's point of view.

**Comment by abadger** [comment]
tldr; If you'd like to write a PR that uses dnf to install python2-dnf and it's acceptable to the maintainers, then that functionality could be added to the module.

Longer explanation:

Ansible is not a package manager; it is a configuration management and orchestration tool.  Ansible is also not a distribution-specific tool; it is a generic tool that runs and manages many different kinds of platforms.  As such, Ansible gives you the ability to bootstrap your deps onto your managed systems and then, from there, proceed to use the rest of the features that you want but it does not pretend to manage the packages that are installed on a system directly.  When a user writes playbooks for their systems they need to be sure that they have playbooks that bootstrap the environment to a state suitable for the modules that they wish to run.  For instance, you installed Python2 onto the Managed system.  That's something that's part of the bootstrapping process for the way you wish to manage your Fedora host.  Installing the Python2 libraries that go along with that is just another part of bootstrapping.  You can do all that configuration of a managed machine from within Ansible if you want using the raw module to install python2 and (if you wish) then using the command module to bootstrap to the point where the python2-dnf module works.

That said, the Ansible core team has discussed in the past whether package management modules are special in that they can assume that the user is specifying to use the "dnf module" because the system in question uses dnf as its primary package manager.  With that assumption, the module could try to invoke the command line tool to do some bootstraping itself.  (The module can install python2-dnf, for instance, but it can't install the python2 interpreter).  We also think there's a line on auto-installation that we don't want to cross (for instance, in the apt module's case, installing aptitude was seen as being too much magic.  For dnf, we wouldn't mind if the package intalls the python2-dnf module but it would be out of scope for it to install the python2 stack itself).  This also does not extend to non-package management modules.  So, for instance, you do need to still install python2 selinux bindings on Fedora to get basic modules like file and copy to work.  These criteria were worked out as a compromise between easier use and the desire by most sysadmins to avoid automagic installation.  They'd rather know that packages are only installed on their systems if they specifically request them.

In response to more of your points:

* Expecting the package module in particular to shield you from platform differences is extremely bad.  The package module is a very light wrapper around each individual package manager module. Only a few of those support autoinstalling deps.  Different platforms have different names for the same packages.  The package module can allow people to run simple playbooks against both Fedora and RHEL OR against both Debian and Ubuntu in some cases but it can't be used to abstract package installation between any two unrelated platforms (for instance, between Debian and Fedora).
* The package modules are not what I'd call "standard".  Most of them are supported by individuals who happen to be using the system on which they are present.  We try to make them support a few common features (like installing packages via a list given via a particular module option name) but we can't control all of their features.  Auto-installation is one that we do not mandate.
* I'm afraid that Ansible works as designed; it's you're expectations that are flawed.  People do need to have custom commands for the specific platforms that they are managing.  Ansible makes it easier to do this by providing you with host facts and conditionals but Ansible does not abstract away all of the differences between systems.
* Ansible does not pull deps for most modules from pypi.  It is up to the sysadmin to install the deps onto the systems that they are managing.  With the exception of package manager modules it is actually impossible for Ansible modules to know how they should pull their deps.  Additionally, in many cases, sysadmins will be upset if a module autoinstalls extra code onto the system.
* Ansible cannot pull the dnf library from pypi because dnf is not present on pypi.
* Ansible is installed onto the controller machine.  Modules are run on remote, managed machines.  installing Ansible can not install deps for the machines that it is managing automatically when Ansible is installed.

**Comment by igor-raits** [comment]
@abadger I meant this part of the code.

```python
def _ensure_dnf(module):
    if not HAS_DNF:
        if PY2:
            package = 'python2-dnf'
        else:
            package = 'python3-dnf'

        if module.check_mode:
            module.fail_json(msg="`{0}` is not installed, but it is required"
                    " for the Ansible dnf module.".format(package))

        module.run_command(['dnf', 'install', '-y', package], check_rc=True)
        global dnf
        try:
            import dnf
            import dnf.cli
            import dnf.const
            import dnf.exceptions
            import dnf.subject
            import dnf.util
        except ImportError:
            module.fail_json(msg="Could not import the dnf python module."
" Please install `{0}` package.".format(package))
```

**Comment by abadger** [comment]
@ignatenkobrain Ah... So it would definitely be wrong for dnf to install packages in check mode.  But it cannot operate without the proper python-dnf package installed beforehand.  I think the checkmode behaviour is less than ideal in this case but it's the best that can be done; users should install the relevant python-dnf package prior to using the dnf module in check-mode (this is the only strategy open to other modules anyway).  You could add a note in the documentation that with check-mode, dnf won't install the python-dnf bindings for the python being run upon.

**Comment by thozza** [comment]
@abadger so what is the issue in my case? AFAIK I'm not running ansible in check mode. I have python2 installed on Fedora and applying the playbook locally on the localhost. I would expect the code to install python2-dnf in my case.

**Comment by thozza** [comment]
tldr; my report was about dnf module not installing python2-dnf package when run, even though it is in its code. IOW dnf module contains a code to install python2-dnf using dnf command, in case python2-dnf is not installed on the system. I expect this is there because someone thought about the actual usefulness of this module. However this is not happening for some reason, supposedly because of a bug. Yet this bug report has been closed because nobody spent time to understand what I actually reported.
